src/server/journal/serializer.cc

```diff
@@ -18,148 +18,186 @@ using namespace std;
   };
-  return std::error_code{};
 }
 
-JournalReader::JournalReader(DbIndex dbid) : buf_{}, dbid_{dbid} {
+JournalReader::JournalReader(io::Source* source, DbIndex dbid)
+    : source_{source}, buf_{}, str_buf_{}, dbid_{dbid} {
+  buf_.EnsureCapacity(4096);
 }
 
 void JournalReader::SetDb(DbIndex dbid) {
   dbid_ = dbid;
 }
 
-template <typename UT> io::Result<UT> ReadPackedUIntTyped(io::Source* source) {
-  uint64_t v;
-  SET_OR_UNEXPECT(ReadPackedUInt(source), v);
-  if (v > std::numeric_limits<UT>::max())
-    return make_unexpected(make_error_code(errc::result_out_of_range));
-  return static_cast<UT>(v);
+void JournalReader::SetSource(io::Source* source) {
+  CHECK_EQ(buf_.InputLen(), 0ULL);
+  source_ = source;
 }
 
-io::Result<uint8_t> JournalReader::ReadU8(io::Source* source) {
-  return ReadPackedUIntTyped<uint8_t>(source);
-}
+std::error_code JournalReader::EnsureRead(size_t num) {
+  // Check if we already have enough.
+  if (buf_.InputLen() >= num)
+    return {};
+
+  uint64_t remainder = num - buf_.InputLen();
+  buf_.EnsureCapacity(remainder);
 
-io::Result<uint16_t> JournalReader::ReadU16(io::Source* source) {
-  return ReadPackedUIntTyped<uint16_t>(source);
+  // Try reading at least how much we need, but possibly more
+  uint64_t read;
+  SET_OR_RETURN(source_->ReadAtLeast(buf_.AppendBuffer(), remainder), read);
+  buf_.CommitWrite(read);
+  return {};
 }
 
-io::Result<uint64_t> JournalReader::ReadU64(io::Source* source) {
-  return ReadPackedUIntTyped<uint64_t>(source);
+template <typename UT> io::Result<UT> JournalReader::ReadUInt() {
+  // Determine type and number of following bytes.
+  EnsureRead(1);
+  PackedUIntMeta meta{buf_.InputBuffer()[0]};
+  buf_.ConsumeInput(1);
+
+  if (auto ec = EnsureRead(meta.ByteSize()); ec)
+    return make_unexpected(ec);
+
+  // Read and check intenger.
+  uint64_t res;
+  SET_OR_UNEXPECT(ReadPackedUInt(meta, buf_.InputBuffer()), res);
+
+  if (res > std::numeric_limits<UT>::max())
+    return make_unexpected(make_error_code(errc::result_out_of_range));
+  return static_cast<UT>(res);
 }
 
-io::Result<size_t> JournalReader::ReadString(io::Source* source) {
+template io::Result<uint8_t> JournalReader::ReadUInt<uint8_t>();
+template io::Result<uint16_t> JournalReader::ReadUInt<uint16_t>();
+template io::Result<uint64_t> JournalReader::ReadUInt<uint64_t>();
+
+io::Result<size_t> JournalReader::ReadString() {
   size_t size = 0;
-  SET_OR_UNEXPECT(ReadU64(source), size);
+  SET_OR_UNEXPECT(ReadUInt<uint64_t>(), size);
 
-  buf_.EnsureCapacity(size);
-  auto dest = buf_.AppendBuffer().first(size);
-  uint64_t read = 0;
-  SET_OR_UNEXPECT(source->Read(dest), read);
+  if (auto ec = EnsureRead(size); ec)
+    return make_unexpected(ec);
 
-  buf_.CommitWrite(read);
-  if (read != size)
-    return make_unexpected(std::make_error_code(std::errc::message_size));
+  unsigned offset = str_buf_.size();
+  str_buf_.resize(offset + size);
+  buf_.ReadAndConsume(size, str_buf_.data() + offset);
```

Comment:
`string::resize` doesn't guarantee me that is resizes with capacity, I probably should to it manually?